draft-ietf-tls-esni.md

```diff
@@ -986,6 +988,24 @@ Offering a GREASE extension is not considered offering an encrypted ClientHello
 for purposes of requirements in {{real-ech}}. In particular, the client
 MAY offer to resume sessions established without ECH.
 
+### Server Greasing
+
+{{config-extensions-iana}} describes a set of Reserved extensions
+which will never be registered. These can be used by servers to
+"grease" the contents of the ECH configuration, as inspired by
+{{?RFC8701}}. This helps ensure clients process ECH extensions
+correctly. When constructing ECH configurations, servers SHOULD
+randomly select from reserved values with the high-order bit
+clear. Correctly-implemented client will ignore those extensions.
+
+The reserved values with the high-order bit set are mandatory, as
+defined in {{config-extensions}}. Servers SHOULD randomly select from
+these values and include them in extraneous ECH configurations.
+Correctly-implemented clients will ignore these configurations because
+they do not recognize the mandatory extension.  These extraneous ECH
+configurations SHOULD have invalid keys, and invalid public names,
```

Comment:
See my comment on #637 for why I am not enthusiastic about this general direction.  I've suggested a different structure there.

Reply:
I don't have a strong opinion about the substance here; this was the previous text.

Reply:
Yeah, I don't think that any of those arguments hold.

For a connection to be established, the server needs to hold the corresponding private key and be willing to use it.  The client needs to use the public name and (somehow) validate it successfully.  The client also needs to ignore the greased codepoint, despite not understanding it and knowing that it means that the entire config should be ignored.  Sure, the client doesn't know that the key pair is or isn't good, but it has multiple cues to go from.

We owe clients that are beyond incompetent no such affordance in specifications.

To take a step back, you only need one clear and unambiguous way to signal that a config should be ignored.  That's the presence of a mandatory, unknown extension.  Adding in name checks a) complicates the validation process and b) prevents me from doing my new, totally awesome thing that might rely on invalid names.  Which I realize that I'm going to have to share now.  Here, it's incomplete, but it might at least make sense: https://github.com/martinthomson/ech-pnmasq

Reply:
I scanned Martin's document and it seems quite complex. I do agree we ought revisit the problem, but *not until after the ECH RFC has issued* so can we rather try get this thing out the door?

Reply:
> Adding in name checks a) complicates the validation process and b) prevents me from doing my new, totally awesome thing ...

The text here is not attempting to add name checks.  In fact, this text assumes that clients do _not_ perform any name checks (beyond syntax).  If client applies special treatment to ".invalid" names, the client has "special-cased the grease" and the procedure is broken.

Would you be happier with an additional sentence like "Clients MUST NOT apply special treatment to names based on the ".invalid" TLD."?